common/src/lib.rs

```diff
@@ -787,4 +803,343 @@ mod tests {
+                program_id,
+                msg_id_2,
+                QueuedDispatch::new_handle(QueuedMessage {
+                    id: msg_id_2,
+                    source: H256::from_low_u64_be(4),
+                    dest: program_id,
+                    payload: Default::default(),
+                    value: 0,
+                    reply: None,
+                }),
+                0,
+            );
+
+            assert_ok!(pause_program(program_id));
+
+            assert!(paused_program_exists(program_id));
+
+            assert!(get_code(code_hash).is_some());
+
+            // although the memory pages should be removed
+            assert_eq!(
+                get_program_pages(program_id, memory_pages.into_keys().collect()),
+                None
+            );
+
+            assert!(remove_waiting_message(program_id, msg_id_1).is_none());
+            assert!(remove_waiting_message(program_id, msg_id_2).is_none());
+        });
+    }
+
+    #[test]
+    fn pause_program_twice_fails() {
+        sp_io::TestExternalities::new_empty().execute_with(|| {
+            let code = b"pretended wasm code".to_vec();
+            let code_hash: H256 = sp_io::hashing::blake2_256(&code[..]).into();
+            set_code(code_hash, &code);
+
+            let program_id = H256::from_low_u64_be(1);
+            let static_pages = 256;
+            set_program(
+                program_id,
+                ActiveProgram {
+                    static_pages,
+                    persistent_pages: Default::default(),
+                    code_hash,
+                    nonce: 0,
+                    state: ProgramState::Initialized,
+                },
+                Default::default(),
+            );
+
+            assert_ok!(pause_program(program_id));
+            assert_err!(pause_program(program_id), PauseError::ProgramNotFound);
+        });
+    }
+
+    #[test]
+    fn pause_terminated_program_fails() {
+        sp_io::TestExternalities::new_empty().execute_with(|| {
+            let code = b"pretended wasm code".to_vec();
+            let code_hash: H256 = sp_io::hashing::blake2_256(&code[..]).into();
+            set_code(code_hash, &code);
+
+            let program_id = H256::from_low_u64_be(1);
+            let static_pages = 256;
+            set_program(
+                program_id,
+                ActiveProgram {
+                    static_pages,
+                    persistent_pages: Default::default(),
+                    code_hash,
+                    nonce: 0,
+                    state: ProgramState::Initialized,
+                },
+                Default::default(),
+            );
+
+            assert_ok!(set_program_terminated_status(program_id));
+
+            assert_err!(pause_program(program_id), PauseError::ProgramTerminated);
+        });
+    }
+
+    fn create_uninitialized_program_messages(
+        static_pages: u32,
+    ) -> (H256, H256, H256, H256, H256, BTreeMap<u32, Vec<u8>>) {
```

Comment:
The function signature is quite ugly, better would use a struct as a return parameter.